src/parse.rs

```diff
@@ -79,7 +79,7 @@ pub(crate) enum ItemBody {
     TaskListMarker(bool), // true for checked
 
     Rule,
-    Heading(HeadingLevel), // heading level
+    Heading(HeadingLevel, HeadingIndex), // heading level
```

Comment:
Would it be faster for headings without id or classes if we'd wrap the `HeadingIndex` in an option? Then we may be able to skip allocations and indirections for that case. It does add another branch, but I suspect it would pay off.

Reply:
Naively using `Option<HeadingIndex>` makes `ItemBody` bigger (24 bytes on x64).

https://github.com/raphlinus/pulldown-cmark/blob/d412d7cb652707f8c980c59dca836a6ad4e49f2f/src/parse.rs#L1495-L1500

It can be easily avoided by using `NonZeroUsize` as the underlying type of `HeadingIndex`, instead of `usize`.
I'm going to modify `HeadingIndex` to use `NonZeroUsize` as I think it is worth doing, but tell me if you think the overhead of using `NonZeroUsize` is unacceptable.